pkg/scheduler/framework/v1alpha1/framework.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"encoding/json"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/klog"
+	internalcache "k8s.io/kubernetes/pkg/scheduler/internal/cache"
+)
+
+// Framework is the component responsible for initializing and
+type Framework struct {
+	registry       Registry
+	pluginContext  *PluginContext
+	cache          *ReadOnlyCache
+	plugins        map[string]Plugin // All plugins
+	reservePlugins []ReservePlugin
+	prebindPlugins []PrebindPlugin
+}
+
+var _ = FrameworkInterface(&Framework{})
+
+// PluginContext returns a pointer to PluginContext that plugins can use to
+// store their temporary data.
+func (f *Framework) PluginContext() *PluginContext {
+	return f.pluginContext
+}
+
+// SchedulerCache returns a pointer to a custom ready-only interface of the
+// scheduler cache.
+func (f *Framework) SchedulerCache() SchedulerCache {
+	return f.cache
+}
+
+// ReservePlugins returns a slice of "reserve" plugins.
+func (f *Framework) ReservePlugins() []ReservePlugin {
+	return f.reservePlugins
+}
+
+// PrebindPlugins returns a slice of "prebind" plugins.
+func (f *Framework) PrebindPlugins() []PrebindPlugin {
+	return f.prebindPlugins
+}
+
+// NewFramework initializes plugins given the configuration and the registry.
+func NewFramework(r Registry, _ json.RawMessage, c internalcache.Cache) *Framework {
+	f := &Framework{
+		registry:      r,
+		pluginContext: NewPluginContext(),
+		cache:         &ReadOnlyCache{cache: c},
+		plugins:       make(map[string]Plugin),
+	}
+
+	// TODO: The framework needs to read the scheduler config and initialize only
+	// needed plugins. In this initial version of the code, we initialize all.
+	for name, factory := range r {
+		// TODO: 'nil' should be replaced by plugin config.
+		p, err := factory(nil, f)
+		if err != nil {
+			klog.Errorf("Error initializing plugin %v: %v", name, err)
+			return nil
```

Comment:
If there is something wrong in initialization, need we return an error and deal with it?

Reply:
+1
If something can go wrong, let's wrap and return the error.

Unless you are worried about making it too easy to for a broken plugin to prevent scheduler from functioning at all? i.e. graceful degradation?